File: src/main/scala/category/heap/_373_Solution.scala
package category.heap

import java.util.{Comparator, PriorityQueue}

/**
  * 373. Find K Pairs with Smallest Sums
  *
  * [1,7,11]
  * [2,4,6]
  * 3
  * [1,1,2]
  * [1,2,3]
  * 2
  * [1,2]
  * [3]
  * 3
  * [1,2]
  * []
  * 2
  * [1,1,1,1]
  * [2,2,2,2,3]
  * 7
  * []
  * []
  * 2
  */
object _373_Solution {
  def kSmallestPairs(nums1: Array[Int], nums2: Array[Int], k: Int): List[List[Int]] = {
    if (nums1.isEmpty || nums2.isEmpty) return Nil
    val heap = new PriorityQueue[Array[Int]](new Comparator[Array[Int]] {
      override def compare(o1: Array[Int], o2: Array[Int]): Int = o1(0) + o1(1) - o2(0) - o2(1)
    })
    for (i <- 0 until Math.min(k, nums1.length)) heap.add(Array(nums1(i), nums2(0), 0))
    var ans: List[List[Int]] = Nil
    for (t <- 0 until k if !heap.isEmpty) {
      val Array(l, r, idx) = heap.poll()
      ans ::= List(l, r)
      if (idx + 1 < nums2.length) heap.add(Array(l, nums2(idx + 1), idx + 1))
    }
    ans.reverse
  }
  
  //    def kSmallestPairs(nums1: Array[Int], nums2: Array[Int], k: Int): List[List[Int]] = {
  //      val heap = new PriorityQueue[(Int, List[Int])](new Comparator[(Int, List[Int])] {
  //        override def compare(o1: (Int, List[Int]), o2: (Int, List[Int])): Int = o1._1 - o2._1
  //      })
  //      val M = nums1.length
  //      val N = nums2.length
  //
  //      def gen: List[List[Int]] = {
  //        var ans: List[List[Int]] = Nil
  //        for (_ <- 0 until k if !heap.isEmpty) {ans ::= heap.poll()._2}
  //        ans.reverse
  //      }
  //
  //      for (i <- 0 until M)
  //        for (j <- 0 until N) {
  //          heap.add((nums1(i) + nums2(j), List(nums1(i), nums2(j))))
  //        }
  //      gen
  //    }
}

File: src/main/python/medium/_300_400/_373_Solution.py
# 373. Find K Pairs with Smallest Sums
import heapq
from typing import List

'''
[1,7,11]
[2,4,6]
3
[1,1,2]
[1,2,3]
2
[1,2]
[3]
4
[]
[1]
2
'''


class _373_Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if not nums1 or not nums2: return []
        heap, ans = [(nums1[0] + nums2[0], 0, 0)], []
        n, m = len(nums1), len(nums2)
        k = min(k, n * m)
        visited = {(0, 0)}
        for _ in range(k):
            _, x, y = heapq.heappop(heap)
            ans.append([nums1[x], nums2[y]])
            if x + 1 < n and (x + 1, y) not in visited:
                heapq.heappush(heap, (nums1[x + 1] + nums2[y], x + 1, y))
                visited.add((x + 1, y))
            if y + 1 < m and (x, y + 1) not in visited:
                heapq.heappush(heap, (nums1[x] + nums2[y + 1], x, y + 1))
                visited.add((x, y + 1))
        return ans

    # full scan
    # def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
    #     if not nums1 or not nums2: return []
    #     heap, ans = [], []
    #     n, m = len(nums1), len(nums2)
    #     k = min(k, n * m)
    #     for i in range(min(n, k)):
    #         for j in range(min(m, k)):
    #             heapq.heappush(heap, (nums1[i] + nums2[j], [nums1[i], nums2[j]]))
    #     for _ in range(k):
    #         ans.append(heapq.heappop(heap)[1])
    #     return ans


if __name__ == '__main__':
    obj = _373_Solution()
    assert obj.kSmallestPairs([1, 1, 2], [1, 2, 3], 10) == [[1, 1], [1, 1], [2, 1], [1, 2], [1, 2], [2, 2], [1, 3],
                                                            [1, 3], [2, 3]]
